GitUI/CommandsDialogs/FormBrowse.cs

```diff
@@ -2818,78 +2822,105 @@ private void SubmoduleStatusProvider_StatusUpdated(object sender, SubmoduleStatu
+                    newItems.Add(CreateSubmoduleMenuItem(result.TopProject, _topProjectModuleFormat.Text));
                 }
 
-                newItems.Add(CreateSubmoduleMenuItem(cancelToken, result.SuperProject, _superprojectModuleFormat.Text));
-                newItems.AddRange(result.AllSubmodules.Select(submodule => CreateSubmoduleMenuItem(cancelToken, submodule)));
+                newItems.Add(CreateSubmoduleMenuItem(result.SuperProject, _superprojectModuleFormat.Text));
+                newItems.AddRange(result.AllSubmodules.Select(submodule => CreateSubmoduleMenuItem(submodule)));
                 toolStripButtonLevelUp.ToolTipText = _goToSuperProject.Text;
             }
 
-            newItems.Add((new ToolStripSeparator(), null));
+            newItems.Add(new ToolStripSeparator());
 
             var mi = new ToolStripMenuItem(updateAllSubmodulesToolStripMenuItem.Text, Images.SubmodulesUpdate);
             mi.Click += UpdateAllSubmodulesToolStripMenuItemClick;
-            newItems.Add((mi, null));
+            newItems.Add(mi);
 
             if (result.CurrentSubmoduleName is not null)
             {
                 var item = new ToolStripMenuItem(_updateCurrentSubmodule.Text) { Tag = result.CurrentSubmoduleName };
                 item.Click += UpdateSubmoduleToolStripMenuItemClick;
-                newItems.Add((item, null));
+                newItems.Add(item);
             }
 
             // Using AddRange is critical: if you used Add to add menu items one at a
             // time, performance would be extremely slow with many submodules (> 100).
-            toolStripButtonLevelUp.DropDownItems.AddRange(newItems.Select(e => e.item).ToArray());
+            toolStripButtonLevelUp.DropDownItems.AddRange(newItems.ToArray());
 
-            // Load details sequentially to not spawn too many background threads
-            // then refresh all items at once with a single switch to the main thread
-            ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+            return newItems;
+        }
+
+        private async Task UpdateSubmoduleMenuStatus(SubmoduleInfoResult result, CancellationToken cancelToken)
+        {
+            if (_currentSubmoduleMenuItems is null)
             {
-                var loadDetails = newItems.Select(e => e.loadDetails).WhereNotNull();
-                var refreshActions = new List<Action>();
-                foreach (var loadFunc in loadDetails)
+                return;
+            }
+
+            await this.SwitchToMainThreadAsync(cancelToken);
+
+            Validates.NotNull(result.TopProject);
+            var infos = result.AllSubmodules.ToDictionary(info => info.Path, info => info);
+            infos[result.TopProject.Path] = result.TopProject;
+            foreach (var item in _currentSubmoduleMenuItems)
+            {
+                var path = (string)item.Tag;
```

Comment:
Use `as` syntax?

Reply:
yes, the next line nullchecks which is what we are after

Reply:
I thought with `as`, it would become shorter and `var` would be saved.